Replace async actions with EC tasks where appropriate

burieberry opened the pull request.

burieberry committed:
replace async actions with EC tasks where appropriate

burieberry committed:
Merge remote-tracking branch 'origin/main' into cs-5833-async-cleanup

burieberry commented:
@habdelra once in a while locally I see the last assertion of this test `can change the item selection in a linksToMany field` fail, so it seems like I'm still missing something. do you see anything obvious?

burieberry commented on packages/host/app/components/operator-mode/overlay-item-header.gts:
```diff
@@ -0,0 +1,161 @@
+      (this.realmInfo = await this.cardService.getRealmInfo(
+        this.args.item.card
+      ))
+  );
```
previously this was: `{{#let (load (this.getRealmInfo card)) as |result|}}...{{/let}}` in the `overlays` template

habdelra replied:
i think since the realmInfo's was using a tracked function (a form of a resource), that this was actually probably ok the way it originally was

burieberry commented on packages/host/app/components/operator-mode/stack-item.gts:
```diff
@@ -146,28 +152,26 @@ export default class OperatorModeStackItem extends Component<Signature> {
+    await navigator.clipboard.writeText(this.card.id);
+  });
 
-  fetchRealmInfo = trackedFunction(this, async () => {
```
not sure if we still need this `trackedFunction`. converted to ember-concurrency

habdelra replied:
actually this is totally cool as a tracked function--a tracked function is a form of a resource. the `this` that is passed into the tracked function binds the lifecycle of the function to the component. I'd suggest moving this back to being a tracked function

habdelra commented on packages/host/app/components/operator-mode/container.gts:
```diff
@@ -570,9 +561,9 @@ export default class OperatorModeContainer extends Component<Signature> {
                 @publicAPI={{this.publicAPI this stackIndex}}
-                @close={{this.close}}
+                @close={{perform this.close}}
                 @edit={{this.edit}}
-                @save={{this.save}}
+                @save={{perform this.save}}
```
nice!

habdelra commented:
> can change the item selection in a linksToMany field

i'm not sure that is related to your PR. i was seeing flakiness there too. I added some additional waits around auto save and thought that fixed it--but maybe not. i can take a  deeper look at this on your branch

burieberry committed:
use tracked function

burieberry committed:
Merge remote-tracking branch 'origin/main' into cs-5833-async-cleanup

burieberry committed:
Merge remote-tracking branch 'origin/main' into cs-5833-async-cleanup

burieberry commented:
> > can change the item selection in a linksToMany field
> 
> i'm not sure that is related to your PR. i was seeing flakiness there too. I added some additional waits around auto save and thought that fixed it--but maybe not. i can take a deeper look at this on your branch

Realized that I removed too many of the `await waitFor('[data-test-save-idle]');`. I'll just remove the ones in the before hook.

habdelra committed:
wait for auto save

habdelra commented:
> > can change the item selection in a linksToMany field
> 
> i'm not sure that is related to your PR. i was seeing flakiness there too. I added some additional waits around auto save and thought that fixed it--but maybe not. i can take a deeper look at this on your branch

ah, so the fix that I'm thinking about I made to a different test. I added that same logic to this test as well. bascially you need to wait for the auto save to complete after each change to the card. I added a test selector `[data-test-last-saved="last saved timespamp"]` that updates each time a save happens. I fixed this test, but I should probably extract this into a nice helper....

habdelra committed:
introduce an auto save helper

burieberry committed:
Merge branch 'cs-5833-async-cleanup' of https://github.com/cardstack/boxel-motion into cs-5833-async-cleanup

habdelra commented:
ok, now I have introduced an auto save helper to make this really easy. so anytime you perform an action that changes the card in some way, just wrap with a `await waitUntilSaved(async () => { do something that changes the card });`

like this:
```    
    await waitFor(`[data-test-stack-card="${testRealmURL}Person/burcu"]`);
    await click('[data-test-edit-button]');

    assert.dom('[data-test-field="friends"]').containsText('Jackie Woody');
    await waitUntilSaved(async () => {
      await click(
        '[data-test-links-to-many="friends"] [data-test-item="1"] [data-test-remove-card]',
      );
    });
    await assert.dom('[data-test-field="friends"]').containsText('Jackie');

    await waitUntilSaved(async () => {
      await click('[data-test-links-to-many="friends"] [data-test-add-new]');
      await waitFor(`[data-test-card-catalog-item="${testRealmURL}Pet/mango"]`);
      await click(`[data-test-select="${testRealmURL}Pet/mango"]`);
      await click('[data-test-card-catalog-go-button]');
    });

    await waitUntil(() => !document.querySelector('[card-catalog-modal]'));
    assert.dom('[data-test-field="friends"]').containsText('Mango');
```

(i should probably make a PR to cleanup the places where I'm waiting the old way to use this new helper)....

habdelra committed:
Merge branch 'cs-5833-async-cleanup' of https://github.com/cardstack/boxel into cs-5833-async-cleanup

habdelra commented:
just FYI, it looks like percy found some serious styling issues with this PR. unsure why it's so different....

burieberry commented:
> just FYI, it looks like percy found some serious styling issues with this PR. unsure why it's so different....

yeah there was a lot of merge conflicts with a recently merged PR. I'll go over it to fix. Once those are fixed, good to merge?

habdelra commented:
> > just FYI, it looks like percy found some serious styling issues with this PR. unsure why it's so different....
> 
> yeah there was a lot of merge conflicts with a recently merged PR. I'll go over it to fix. Once those are fixed, good to merge?

yup!

habdelra commented:
> > > can change the item selection in a linksToMany field
> > 
> > 
> > i'm not sure that is related to your PR. i was seeing flakiness there too. I added some additional waits around auto save and thought that fixed it--but maybe not. i can take a deeper look at this on your branch
> 
> Realized that I removed too many of the `await waitFor('[data-test-save-idle]');`. I'll just remove the ones in the before hook.

that beforeHook's data-test-save-idle is not horrible to have in there (it's definitely useful for you to remove it to see if you have done things correctly). generally when things are working it is unnecessary, but for broken tests it's probably really helpful in that it will help to remove errors that are red herrings.

burieberry committed:
missing div and changes from prettier

burieberry committed:
revert change in hook
